runtime/src/committee.rs

```diff
@@ -0,0 +1,716 @@
+pub type Origin<T, I = DefaultInstance> = RawOrigin<<T as system::Trait>::AccountId, I>;
+
+#[derive(PartialEq, Eq, Clone, Encode, Decode, Debug)]
+/// Info for keeping track of a motion being voted on.
+pub struct Votes<IdentityId> {
+    /// The proposal's unique index.
+    index: ProposalIndex,
+    /// The current set of commmittee members that approved it.
+    ayes: Vec<IdentityId>,
+    /// The current set of commmittee members that rejected it.
+    nays: Vec<IdentityId>,
+}
+
+decl_storage! {
+    trait Store for Module<T: Trait<I>, I: Instance=DefaultInstance> as Committee {
+        /// The hashes of the active proposals.
+        pub Proposals get(proposals): Vec<T::Hash>;
+        /// Actual proposal for a given hash.
+        pub ProposalOf get(proposal_of): map T::Hash => Option<<T as Trait<I>>::Proposal>;
+        /// Votes on a given proposal, if it is ongoing.
+        pub Voting get(voting): map T::Hash => Option<Votes<IdentityId>>;
+        /// Proposals so far.
+        pub ProposalCount get(proposal_count): u32;
+        /// The current members of the committee.
+        pub Members get(members) config(): Vec<IdentityId>;
+        /// Vote threshold for an approval.
+        pub VoteThreshold get(vote_threshold) config(): (ProportionMatch, u32, u32);
+    }
+    add_extra_genesis {
+        config(phantom): rstd::marker::PhantomData<(T, I)>;
+    }
+}
+
+decl_event!(
+	pub enum Event<T, I=DefaultInstance> where
+		<T as system::Trait>::Hash,
+	{
+		/// A motion (given hash) has been proposed (by given account) with a threshold (given
+		/// `MemberCount`).
+		Proposed(IdentityId, ProposalIndex, Hash),
+		/// A motion (given hash) has been voted on by given account, leaving
+		/// a tally (yes votes, no votes and total seats given respectively as `MemberCount`).
+		Voted(IdentityId, Hash, bool, MemberCount, MemberCount, MemberCount),
+		/// A motion was approved by the required threshold with the following
+		/// tally (yes votes, no votes and total seats given respectively as `MemberCount`).
+		Approved(Hash, MemberCount, MemberCount, MemberCount),
+		/// A motion was rejected by the required threshold with the following
+		/// tally (yes votes, no votes and total seats given respectively as `MemberCount`).
+		Rejected(Hash, MemberCount, MemberCount, MemberCount),
+		/// A motion was executed; `bool` is true if returned without error.
+		Executed(Hash, bool),
+	}
+);
+
+decl_module! {
+    pub struct Module<T: Trait<I>, I: Instance=DefaultInstance> for enum Call where origin: <T as system::Trait>::Origin {
+        fn deposit_event() = default;
+
+        /// Set the committee's membership manually to `new_members`.
+        /// Requires root origin.
+        ///
+        /// # Arguments
+        /// * `origin` Root
+        /// * `new_members` Members to be initialized as committee.
+        #[weight = SimpleDispatchInfo::FixedOperational(100_000)]
+        fn set_members(origin, new_members: Vec<IdentityId>) {
```

Comment:
Do we not also need to either manage the membership identities via group.rs or add to this module specific functions to add / remove members (which have an origin that corresponds to a majority of existing members)?

Reply:
`set_members` is just for initializing the module. 

To change membership, we could do similar to what Kusama does -- use an instance of `group` to add/remove members. The `group` instance would be controlled by a different instance of `committee`. So, in total there would be 2 `committee`s and 1 `group` in runtime. The downside is there's a new committee to add/remove members of `group` which holds `committee` members.

Reply:
`set_members` is not needed for the initialization but is helpful for testing. We can simply remove it (like other root functions) before going live.

As far as adding/removing members go, it makes sense for the `group` module to manage that but I don't understand the need of two `committee`. Wouldn't the original/target `committee` be able to control the `group`?

Reply:
If this is just for testing, we can add `#[cfg(test)]` before `set_members` function.
In that case we will be sure that we won't forget it when we go to production code ;)

Reply:
From testing, I meant UI testing. For that, we need this function to be part of release build so `#[cfg(test)]` won't work for us here. It's fine though. Even if we forget to remove this, the only thing that will happen is that nobody will be able to call this function.

Reply:
We decided to lease `set_members` to initialize the committee. 

I've added `CommitteeMembership` (an instance of `group`) to manage members. `PolymeshCommittee` is the governance commitee. Members of which propose and vote to add/remove members.

Reply:
Eventually, calls relying on root will be removed in favor of governance process.